File: 2-TLC/SCRAMBLING/Scrambling.cpp

/*************************************************************************
 *** Title: ADDITIVE AND MULTIPLICATIVE SCRAMBLING
 *** Author: Filippo Valmori
 *** Date: 16/10/2018
 *** Reference: [1] Wikipedia - https://en.wikipedia.org/wiki/Scrambler
 *************************************************************************/


/*** NEEDED LIBRARIES ***/

#include <stdio.h>														// Standard library
#include <stdint.h>														// To use well-defined integer types



/*** PARAMETERS ***/

const uint8_t NcA = 15;													// Number of additive LFSR cells
const uint32_t ConVectA = 0x00060000; 									// Connection vector of additive LFSRs, i.e. (1+)z^-14+z^-15 - DVB standard
const uint32_t InitStA = 0x95000000;									// Initial state of additive LFSRs

const uint8_t NcM = 17;													// Number of multiplicative LFSR cells
const uint32_t ConVectM = 0x00108000; 									// Connection vector of multiplicative LFSRs, i.e. (1+)z^-12+z^-17
const uint32_t InitStM = 0x2C350000;									// Initial state of the multiplicative LFSRs



/*** FUNCTION PROTOTYPES ***/

void AddScramb( uint8_t *IoBytes, uint16_t ByteLen );
void MultScramb( uint8_t *IoBytes, uint16_t ByteLen );
void MultDescramb( uint8_t *IoBytes, uint16_t ByteLen );
uint8_t ComputeRegBit( uint32_t State, uint8_t Nc, uint32_t ConVect );

void PrintArray( uint8_t *InBytes, uint16_t InLen );



/*** MAIN FUNCTION ***/

int main(){
	
	uint16_t Len = 50;													// Info stream length (in bytes)
	uint8_t ByteStream[Len] = {0};										// All-zero info stream generation
	
	printf("\n * Scrambling type : Additive");
	printf("\n * Number of LFSR cells = %d",NcA);
	printf("\n * Info stream length = %d bytes\n",Len);
	printf("\n INF :\t");
	PrintArray(ByteStream,Len);											// Print the original info bytes
	AddScramb(ByteStream,Len);											// Apply additive scrambling
	printf("\n SCR :\t");
	PrintArray(ByteStream,Len);											// Print the scrambled bytes
	AddScramb(ByteStream,Len);											// Apply additive descrambling
	printf("\n DES :\t");
	PrintArray(ByteStream,Len);											// Print the descrambled bytes

	printf("\n       --------------------------------\n");

	printf("\n * Scrambling type : Multiplicative");
	printf("\n * Number of LFSR cells = %d",NcM);
	printf("\n * Info stream length = %d bytes\n",Len);
	printf("\n INF :\t");
	PrintArray(ByteStream,Len);											// Print the original info bytes
	MultScramb(ByteStream,Len);											// Apply multiplicative scrambling
	printf("\n SCR :\t");
	PrintArray(ByteStream,Len);											// Print the scrambled bytes
	MultDescramb(ByteStream,Len);										// Apply multiplicative descrambling
	printf("\n DES :\t");
	PrintArray(ByteStream,Len);											// Print the descrambled bytes


}



/*** SCRAMBLING-DESCRAMBLING FUNCTIONS ***/

/**
 *	Function for implementing an additive (aka synchronous) scrambler / descrambler.
 **/
void AddScramb( uint8_t *IoBytes, uint16_t ByteLen ){
	
	uint16_t j,ByteIdx;
	uint8_t BitIdx, ScrBit;
	uint8_t Mask = 0x01;
	uint32_t LfsrState = InitStA;
	uint16_t BitLen = (ByteLen<<3);
	
	for ( j=0; j<BitLen; j++ ){
		
		ByteIdx = (j>>3);
		BitIdx = (uint8_t)(7-(j&0x0007));
		ScrBit = ComputeRegBit(LfsrState,NcA,ConVectA);					// Bit outgoing from shift register connections
		LfsrState >>= 1;												// Update scrambler shift register
		
		if( ScrBit ){													// If LFSR bit is 1
		
			LfsrState |= 0x80000000;									// Update content of shift register first cell
			ScrBit ^= (IoBytes[ByteIdx]>>BitIdx) & Mask;				// Output bit at j-th cycle
			
			if ( ScrBit )												// If Output bit is 1
				IoBytes[ByteIdx] |= (Mask<<BitIdx);						// Write the the output bit into the I/O stream
			else														// If Output bit is 0
				IoBytes[ByteIdx] &= ~(Mask<<BitIdx);
			
		} else {														// If LFSR bit is 0
		
			LfsrState &= 0x7FFFFFFF;
			ScrBit ^= (IoBytes[ByteIdx]>>BitIdx) & Mask;
			
			if ( ScrBit )
				IoBytes[ByteIdx] |= (Mask<<BitIdx);
			else
				IoBytes[ByteIdx] &= ~(Mask<<BitIdx);
			
		}
	}
	
}


/**
 *	Function for implementing a multiplicative (aka self-synchronizing) scrambler.
 **/
void MultScramb( uint8_t *IoBytes, uint16_t ByteLen ){
	
	uint16_t j,ByteIdx;
	uint8_t BitIdx, OutBit, InBit;
	uint8_t Mask = 0x01;
	uint32_t LfsrState = InitStM;
	uint16_t BitLen = (ByteLen<<3);
	
	for ( j=0; j<BitLen; j++ ){
		
		ByteIdx = (j>>3);
		BitIdx = (uint8_t)(7-(j&0x0007));
		InBit = ((IoBytes[ByteIdx]>>BitIdx) & Mask);					// Input bit at j-th cycle
		OutBit = ComputeRegBit(LfsrState,NcM,ConVectM)^InBit;			// Output bit at j-th cycle
		LfsrState >>= 1;												// Right-shift the LFSR
		
		if( OutBit ){													// If output bit is 1
		
			LfsrState |= 0x80000000;									// Update content of LFSR first cell to 1
			IoBytes[ByteIdx] |= (Mask<<BitIdx);							// Write the output bit into the I/O stream

		} else {														// If LFSR bit is 0
		
			LfsrState &= 0x7FFFFFFF;									// Update content of LFSR first cell to 0
			IoBytes[ByteIdx] &= ~(Mask<<BitIdx);
			
		}
	}
	
}


/**
 *	Function for implementing a multiplicative (aka self-synchronizing) descrambler.
 **/
void MultDescramb( uint8_t *IoBytes, uint16_t ByteLen ){
	
	uint16_t j,ByteIdx;
	uint8_t BitIdx, OutBit, InBit;
	uint8_t Mask = 0x01;
	uint32_t LfsrState = InitStM;
	uint16_t BitLen = (ByteLen<<3);
	
	for ( j=0; j<BitLen; j++ ){
		
		ByteIdx = (j>>3);
		BitIdx = (uint8_t)(7-(j&0x0007));
		InBit = ((IoBytes[ByteIdx]>>BitIdx) & Mask);
		OutBit = ComputeRegBit(LfsrState,NcM,ConVectM)^InBit;			// Output bit at j-th cycle
		LfsrState >>= 1;												// Right-shift the LFSR

		if ( OutBit ){													// If output bit is 1
			IoBytes[ByteIdx] |= (Mask<<BitIdx);							// Write the output bit into the I/O stream
		} else {
			IoBytes[ByteIdx] &= ~(Mask<<BitIdx);
		}

		if ( InBit ){
			LfsrState |= 0x80000000;									// Update content of the LFSR first cell to 1
		} else {
			LfsrState &= 0x7FFFFFFF;									// Update content of the LFSR first cell to 0
		}
	}
	
}


/**
 *	Function for computing the bit outgoing from shift register connections
 *  as a function of the current state and connection vector.
 **/
uint8_t ComputeRegBit( uint32_t State, uint8_t Nc, uint32_t ConVect ){

	uint8_t j;
	uint8_t OutBit = 0;
	uint32_t Mask = 0x80000000;

	for ( j=0; j<Nc; j++ ){
		if ( ((State<<j) & Mask) & ((ConVect<<j) & Mask) ){
			OutBit++;
		}
	}
	
	return (OutBit%2);

}



/*** EXTRA FUNCTIONS ***/

/**
 *	Function for printing on terminal (in hexadecimal format) the input byte-array content.
 **/
void PrintArray( uint8_t *InBytes, uint16_t InLen ){

	uint16_t j;
	for ( j=0; j<InLen; j++){
		printf("%2X ",InBytes[j]);
		if ( (j%10 == 9) && (j<(InLen-1)) ) {
			printf("\n\t");

		}
	}
	printf("\n");

}

File: 2-TLC/SCRAMBLING/Scrambling.m

%%% Title: ADDITIVE / MULTIPLICATIVE SCRAMBLING AND PSEUDO-RANDOM GENERATOR
%%% Author: Filippo Valmori
%%% Date: 13/10/2018
%%% Reference: [1] Wikipedia - https://en.wikipedia.org/wiki/Scrambler

close all; clearvars; clc


%% SIMULATION #1 - ADDITIVE SCRAMBLER

% InfoLen = 600;                                                      % Number of source info bits
% Nc = 15;                                                            % Number of scrambler/descramler LFSR cells
% ConVect = [0 0 0 0 0 0 0 0 0 0 0 0 0 1 1];                          % Scrambler/descramler connection vector, i.e. (1+)z^-14+z^-15 - DVB standard
% InitStateTx = [1 0 0 1 0 1 0 1 0 0 0 0 0 0 0];                      % Scrambler LFSR initial state
% InitStateRx = [1 0 0 1 0 1 0 1 0 0 0 0 0 0 0];                      % Descrambler LFSR initial state
% Peb = 0e-2;                                                         % Error probabilty over Binary Symmetric Channel (BSC)
% 
% InfoStream = randi([0 1],1,InfoLen);                                % Random generation of source info bits
% CheckParam(Nc,ConVect,InitStateTx,InitStateRx);                     % Check the parameters correctness
% ScrStream = AddScramb(InfoStream,ConVect,InitStateTx);              % Additive scrambled stream
% ChanStream = bsc(ScrStream,Peb);                                    % BSC simulation
% DesStream = AddScramb(ChanStream,ConVect,InitStateRx);              % Additive descrambled stream
% 
% fprintf(" PARAMETERS:\n");
% fprintf(" - Scrambling type : Additive\n")
% fprintf(" - Info stream length = %d bits\n",InfoLen);
% fprintf(" - Number of LFSR cells = %d\n",Nc);
% fprintf(" - Tx/Rx LFSR initial state mismatch = %d bits\n", ...
%     sum(InitStateTx~=InitStateRx));
% fprintf(" - Probability of error over BSC = %1.3f\n",Peb);
% fprintf("\n RESULTS:\n");
% fprintf(" - Errors before descrambler = %d out of %d bits\n", ...
%     sum(ScrStream~=ChanStream),InfoLen);
% fprintf(" - Errors after descrambler = %d out of %d bits\n\n", ...
%     sum(InfoStream~=DesStream),InfoLen);



%% SIMULATION #2 - MULTIPLICATIVE SCRAMBLER

% InfoLen = 600;                                                      % Number of source info bits
% Nc = 17;                                                            % Number of scrambler/descramler LFSR cells
% ConVect =  [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1];                     % Scrambler/descramler connection vector, i.e. (1+)z^-12+z^-17 - AX.25 standard
% InitStateTx = [0 0 1 0 1 1 0 0 0 0 1 1 0 1 0 1 0];                  % Scrambler LFSR initial state
% InitStateRx = [1 1 1 0 1 1 0 1 1 0 0 1 1 1 1 1 1];                  % Descrambler LFSR initial state
% Peb = 0e-2;                                                         % Error probabilty over Binary Symmetric Channel (BSC)
% 
% InfoStream = randi([0 1],1,InfoLen);                                % Random generation of source info bits
% CheckParam(Nc,ConVect,InitStateTx,InitStateRx);                     % Check the parameters correctness
% ScrStream = MultScramb(InfoStream,ConVect,InitStateTx);             % Multiplicative scrambled stream
% ChanStream = bsc(ScrStream,Peb);                                    % BSC simulation
% DesStream = MultDescramb(ChanStream,ConVect,InitStateRx);           % Multiplicative descrambled stream
% 
% fprintf("     -------------------------------------\n\n");
% fprintf(" PARAMETERS:\n");
% fprintf(" - Scrambling type : Multiplicative\n")
% fprintf(" - Info stream length = %d bits\n",InfoLen);
% fprintf(" - Number of LFSR cells = %d\n",Nc);
% fprintf(" - Tx/Rx LFSR initial state mismatch = %d bits\n", ...
%     sum(InitStateTx~=InitStateRx));
% fprintf(" - Probability of error over BSC = %1.3f\n",Peb);
% fprintf("\n RESULTS:\n");
% fprintf(" - Errors before descrambler = %d out of %d bits\n", ...
%     sum(ScrStream~=ChanStream),InfoLen);
% fprintf(" - Errors after descrambler = \n\t");
% fprintf(" %d out of %d bits (before sync)\n\t", ...
%     sum(InfoStream(1:Nc)~=DesStream(1:Nc)),Nc);
% fprintf(" %d out of %d bits (after sync)\n", ...
%     sum(InfoStream(Nc+1:end)~=DesStream(Nc+1:end)),InfoLen-Nc);



%% SIMULATION #3 - ADDITIVE SCRAMBLER AS PSEUDO-RANDOM GENERATOR

Nc = 5;
InfoLen = 2*(2^Nc-1)+5;                                             % Choosen long enough to create two periods of the generated PN sequence

if Nc == 3
    ConVect = [0 1 1];
elseif Nc == 4
    ConVect = [0 0 1 1];
elseif Nc == 5
    ConVect = [0 0 1 0 1];
else
    Nc = 6;
    ConVect = [0 0 0 0 1 1];
end

InitSt = [1 randi([0 1],1,Nc-1)];                                   % Since the input sequence is all-zero, it must be assured that at least one bit of the initial state is '1'

InfoStream = zeros(InfoLen,1);
ScrStream = AddScramb(InfoStream,ConVect,InitSt);

fprintf(" * Scrambler degree / number of cells = %d\n",Nc);
fprintf(" * Pseudo-noise sequence period = %d bits\n",2^Nc-1);


figure; box on; hold on
Time = 0:length(ScrStream)-1;
plot(Time,ScrStream,'b.-')
line(2^Nc-1*[1 1],[-.2 1.2],'LineWidth',1,'Color',[1 .2 0],'LineStyle','-.')
line(2*(2^Nc-1)*[1 1],[-.2 1.2],'LineWidth',1,'Color',[1 .2 0],'LineStyle','-.')
ylim([-.5 1.5]); grid on
title(sprintf("PSEUDO-NOISE SEQUENCE | Nc = %d",Nc))
hold off



%% FUNCTIONS

%%% --> Function for checking the parameters correctness
function [ ] = CheckParam( Ncells, ConVect, InitStateTx, InitStateRx )

    if length(ConVect) ~= Ncells
        error('Invalid connection vector');
    elseif length(InitStateTx) ~= Ncells
        error('Invalid Tx initial state');
    elseif length(InitStateRx) ~= Ncells
        error('Invalid Rx initial state');
    end
    
end


%%% Function for implementing an additive scrambler (aka synchronous or
%%% linear-feedback shift register scrambler) and descrambler
function [OutBits] = AddScramb( InBits, ConVect, InitState )

    BitLen = length(InBits);
    RegState = InitState;
    OutBits = zeros(1,BitLen);
    
    for j = 1:BitLen
        
        RegBit = mod(sum(RegState.*ConVect),2);
        OutBits(j) = xor(InBits(j),RegBit);
        RegState(2:end) = RegState(1:end-1);
        RegState(1) = RegBit;
        
    end

end


%%% --> Function for implementing a multiplicative scrambler at bit level
function [ OutBits ] = MultScramb( InBits, ConVect, InitState )

    BitLen = length(InBits);
    RegState = InitState;
    OutBits = zeros(1,BitLen);
    
    for j = 1:BitLen
        
        RegBit = mod(sum(RegState.*ConVect),2);
        OutBits(j) = xor(InBits(j),RegBit);
        RegState(2:end) = RegState(1:end-1);
        RegState(1) = OutBits(j);
        
    end

end


%%% --> Function for implementing a multiplicative descrambler at bit level
function [ OutBits ] = MultDescramb( InBits, ConVect, InitState )

    BitLen = length(InBits);
    RegState = InitState;
    OutBits = zeros(1,BitLen);
    
    for j = 1:BitLen
        
        RegBit = mod(sum(RegState.*ConVect),2);
        OutBits(j) = xor(InBits(j),RegBit);
        RegState(2:end) = RegState(1:end-1);
        RegState(1) = InBits(j);
        
    end

end



%% NOTES

% NB#1: Keep in mind that the last element of ConVect must be alway set to '1', otherwise the scrambler degree would be lower 
% than expected.
% NB#2: In SIMLATION#3 by using an all-one info stream (instead of an all-zero as proposed), the output PN sequence would be
% bit-wise XORed.
% NB#3: In SIMLATION#3 ConVect and InitSt shall be choosen accuretely, otherwise the output PN sequence may have a period
% shorter than expected.
